server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,314 @@
+            except IndexError:
+                break
+        self._logger.debug("Rating disparity: %s", elem1.rating)
+
+        # We now need to open all containers again to get to the searches. A container can hold a single
+        # search or two other containers, so we differentiate the two cases by the length of the content array.
+        # Because each container represent the difference of the two containing elements they have to go in
+        # different teams (The higher one into the team the container is in).
+        team_a = []
+        team_b = []
+        containers_a = []
+        containers_b = []
+        containers_a.append(elem1.content[0])
+        containers_b.append(elem1.content[1])
+        while containers_a or containers_b:
+            for e in containers_a:
+                if len(e.content) == 2:
+                    containers_a.append(e.content[0])
+                    containers_b.append(e.content[1])
+                else:
+                    team_a.append(e.content[0])
+                containers_a.remove(e)
+            for e in containers_b:
+                if len(e.content) == 2:
+                    containers_b.append(e.content[0])
+                    containers_a.append(e.content[1])
+                else:
+                    team_b.append(e.content[0])
+                containers_b.remove(e)
+        return team_a, team_b
+
+    def _searches_by_size(self, searches: List[Search]) -> Dict[int, List[Search]]:
+        searches_by_size: Dict[int, List[Search]] = defaultdict(list)
+
+        for search in searches:
+            searches_by_size[len(search.players)].append(search)
+
+        if self._logger.isEnabledFor(logging.DEBUG):
+            max_size = max(searches_by_size.keys())
+            self._logger.debug("participating searches by player size:")
+            for i in range(1, max_size + 1):
+                self._logger.debug("%i players: %s" % (i, searches_by_size[i]))
+        return searches_by_size
+
+    def _find_most_balanced_filler(self, avg, search, searches_dict):
+        """
+        If we simply fetch the highest/lowest rated single player search we may overshoot our
+        goal to get the most balanced teams, so we try them all to find the one that brings us
+        closest to the rating average i.e. balanced teams
+        If there is no single player search we have hit a search combination that is impossible to
+        separate into two teams e.g. (3, 3, 2) for 4v4
+        """
+        if not searches_dict[1]:
+            self._logger.warning("given searches are impossible to split in even teams because of party sizes")
+            raise UnevenTeamsException()
+
+        iterator = iter(searches_dict[1])
+        candidate = next(iterator)
+        old_team_avg = get_average_rating([search, candidate])
+        old_avg_delta = abs(avg - old_team_avg)
+        self._logger.debug("delta with %s is %s (avg is %s)", [candidate], old_avg_delta, old_team_avg)
```

Comment:
You could simplify this duplication by just setting `old_avg_delta` to `math.inf` here. I'd probably also rename it to `min_delta` since that's what you're trying to find.

I think this this whole function could also be a one liner (only you will lose the logging):
```python3
return min(
    searches_dict[1], 
    key=lambda item: abs(avg - get_average_rating([search, item]))
)
```

Reply:
lol, you are right, I didn't even notice I basically reimplemented the min function :D